### ISSUE TITLE
[社区任务4] 基于megatron+vllm_ascend完成qwen2.5 0.5b的 CISPO训练

### ISSUE THREAD

**Issue by wucong25** [verl-project/verl-ascend-recipe#17]
- 任务描述
"昇腾上基于VeRL开放仓库的vllm_ascend推理后端，补齐该模型CISPO训练对于megatron后端的支持。
本期任务的具体信息如下：
（1）模型名称与算法：qwen2.5 0.5b+CISPO
（2）VeRL仓库的模型源码链接：
https://github.com/verl-project/verl
（3）参考脚本：examples\cispo_trainer\run_cispo_qwen2_5_0_5b_gsm8k.sh
（4） 开发平台：Atlas 800T A2或者A3"

- 验收标准
""一、任务交付件
本期任务为基于verl开放仓库代码进行算法扩展，请合入开发代码。主要开发点如下：
（1） 模型适配patch
（2） 算法优化实现
二、验收标准：
1）精度/性能要求
精度：
长跑12小时或者训练100步，reward上升，平均误差小于5%，下游评测误差5%内。如果无GPU标杆，reward上升即可。
性能：
A2上达成A100的0.7倍以上。如果无GPU标杆，TPS绝对值大于100即可。
2）实践文档：
算法适配调优文档1篇
3）任务完成标准
本次任务完成标准为：
精度/性能（根据实际要求）达标，PR完成合入，实践文档提交到仓库issue。""

- 任务类型
算法支持

**Comment by wangyam1** [comment]
认领这个任务

**Comment by Ronald1995** [comment]
> 认领这个任务

@wangyam1 
欢迎认领任务，请参考https://github.com/verl-project/verl/issues/6439  社区任务池明确该任务的：

- 完成的截止日期
- 开发进展反馈
- 微信答疑群
- 任务交付注意事项

等信息。如果您同时认领了多项任务，但无法都能进行投入，可以在部分任务中回复退出.

麻烦您加入到对应微信群，群备注名修改为"社区任务+您的gitcode账号", 后续有相关消息和问题都可以在微信群咨询答疑。 等您加入到微信群后，我这边会在社区任务池里面登记任务责任人。

**Comment by jianzhnie** [comment]
认领这个任务

**Comment by chengminhua** [comment]
> 认领这个任务

欢迎参加认领社区任务，当前一个任务只分配一个开发，根据评论认领时间确定，您可以后续观察该任务是否被释放退出，或者认领其他任务 https://gitcode.com/org/Ascend/discussions/4

**Comment by chengminhua** [comment]
> 认领这个任务

你好，请关注下https://github.com/verl-project/verl/issues/6439 里的任务规则，每周一更新下进展哦

**Comment by chengminhua** [comment]
> 认领这个任务

你好，请关注下https://github.com/verl-project/verl/issues/6439 里的任务规则，每周一更新下进展哦

**Comment by duanzhaol** [comment]
认领这个任务

**Comment by chengminhua** [comment]
> 认领这个任务

感谢您对昇腾开源社区的关注与支持！
由于该任务在认领后较长时间内暂未收到进展更新，为保障项目整体推进节奏，我们将暂时结束本次认领，并重新开放该任务，供其他社区贡献者参与。
后续如您仍有兴趣参与昇腾开源项目，欢迎持续关注【昇腾生态官网】及本社区发布的最新任务，也欢迎通过提交 Issue或参与社区讨论等方式继续贡献。
再次感谢您的理解与支持，期待未来继续与您合作！

**Comment by chengminhua** [comment]
> 认领这个任务

感谢您此前对该任务的关注与支持！
由于上一位认领者暂未继续推进该任务，目前该任务已重新开放。根据此前的候选顺序，我们希望优先邀请您参与该任务。
如您目前仍有意愿认领，请在5个工作日内回复确认，并简要说明预计开始时间或初步推进计划。确认后，我们将为您保留任务，并提供必要的背景资料和支持。
如您近期时间安排不便，也请直接告知，我们将继续顺延联系其他候选贡献者，不会影响您后续参与其他社区任务。
再次感谢您对昇腾开源社区的关注，期待与您合作！

**Comment by RordChang** [comment]
认领这个任务

**Comment by chengminhua** [comment]
> 认领这个任务

感谢您此前对该任务的关注与支持！
由于上一位认领者暂未继续推进该任务，目前该任务已重新开放。根据此前的候选顺序，我们希望优先邀请您参与该任务。
如您目前仍有意愿认领，请在5个工作日内回复确认，并简要说明预计开始时间或初步推进计划。确认后，我们将为您保留任务，并提供必要的背景资料和支持。
如您近期时间安排不便，也请直接告知，我们将继续顺延联系其他候选贡献者，不会影响您后续参与其他社区任务。
再次感谢您对昇腾开源社区的关注，期待与您合作！

**Comment by RordChang** [comment]
> > 认领这个任务
> 
> 感谢您此前对该任务的关注与支持！ 由于上一位认领者暂未继续推进该任务，目前该任务已重新开放。根据此前的候选顺序，我们希望优先邀请您参与该任务。 如您目前仍有意愿认领，请在5个工作日内回复确认，并简要说明预计开始时间或初步推进计划。确认后，我们将为您保留任务，并提供必要的背景资料和支持。 如您近期时间安排不便，也请直接告知，我们将继续顺延联系其他候选贡献者，不会影响您后续参与其他社区任务。 再次感谢您对昇腾开源社区的关注，期待与您合作！

确认认领

**Comment by OnPathXD** [comment]
认领这个任务

**Comment by fangweii1** [comment]
> > > 认领这个任务
> > 
> > 
> > 感谢您此前对该任务的关注与支持！ 由于上一位认领者暂未继续推进该任务，目前该任务已重新开放。根据此前的候选顺序，我们希望优先邀请您参与该任务。 如您目前仍有意愿认领，请在5个工作日内回复确认，并简要说明预计开始时间或初步推进计划。确认后，我们将为您保留任务，并提供必要的背景资料和支持。 如您近期时间安排不便，也请直接告知，我们将继续顺延联系其他候选贡献者，不会影响您后续参与其他社区任务。 再次感谢您对昇腾开源社区的关注，期待与您合作！
> 
> 确认认领

你好，请关注下https://github.com/verl-project/verl/issues/6439 里的任务规则，每周一更新下进展哦

**Comment by RordChang** [comment]
已经跑完训练任务，正在整理文档

**Comment by RordChang** [comment]
## CISPO Megatron + vLLM-Ascend 设计文档与阶段结果

### 1. 目标与范围

本任务为 Qwen2.5-0.5B-Instruct 补齐 CISPO 在 Ascend NPU 上的 Megatron 训练与 vLLM-Ascend rollout 示例。训练侧的 actor/reference 使用 Megatron，采样侧使用 vLLM-Ascend；数据集为 GSM8K，完成 100-step 训练并记录 reward 与 TPS。

主仓 PR： https://github.com/verl-project/verl/pull/7442
Recipe仓 PR:  https://github.com/verl-project/verl-ascend-recipe/pull/85
### 2. 算法与调用链

CISPO 使用当前策略与旧策略 token log probability 的 importance ratio，并对截断后的 ratio 停止梯度：

```text
ratio = exp(log_pi - log_pi_old)
clipped_ratio = stop_gradient(clip(ratio, 1 - eps_low, 1 + eps_high))
loss = -mean_masked(clipped_ratio * advantage * log_pi)
```

本次配置为：

```text
algorithm.adv_estimator=grpo
actor_rollout_ref.actor.policy_loss.loss_mode=cispo
actor_rollout_ref.actor.clip_ratio_low=10.0
actor_rollout_ref.actor.clip_ratio_high=0.2
actor_rollout_ref.actor.loss_agg_mode=token-mean
```

完整训练链路：

```text
GSM8K parquet
  -> vLLM-Ascend rollout
  -> GSM8K exact-answer reward
  -> GRPO group advantage
  -> CISPO token objective
  -> Megatron actor update
  -> bucketed weight synchronization
  -> next rollout
```

CISPO objective 通过 verl 已注册的统一 policy-loss 调用链执行，Megatron actor 直接复用该实现。

### 3. 主仓改动

- `examples/cispo_trainer/run_qwen2_5_0_5b_megatron.sh`
  - Qwen2.5-0.5B-Instruct + GSM8K；
  - Megatron actor/reference，TP=2、PP=1、BF16、MBridge；
  - vLLM-Ascend rollout，TP=2、每个 prompt 采样 4 个 response；
  - 默认 4 NPU、100 steps；
  - 模型、数据、并行度、batch、checkpoint 与日志目录均可通过环境变量覆盖；
  - `resume_mode=auto`，日志按时间戳保存。
- `examples/cispo_trainer/README.md`
  - 增加简洁入口命令与任务结果链接。
- `tests/trainer/ppo/test_dynamic_policy_losses_on_cpu.py`
  - 覆盖 CISPO ratio 截断、stop-gradient、response mask 和 policy-loss registry。

### 4. 数据准备与运行

```bash
python3 examples/data_preprocess/gsm8k.py \
    --local_save_dir /path/to/data/gsm8k

MODEL_PATH=/path/to/Qwen2.5-0.5B-Instruct \
DATA_ROOT=/path/to/data \
NPUS_PER_NODE=4 \
TOTAL_TRAINING_STEPS=100 \
bash examples/cispo_trainer/run_qwen2_5_0_5b_megatron.sh
```

默认 checkpoint 位于：

```text
./checkpoints/qwen2_5_0_5b_cispo_megatron_vllm_ascend/
```

默认训练日志位于：

```text
./logs/qwen2_5_0_5b_cispo_megatron_vllm_ascend_<timestamp>.log
```

容器需要为配置的 weight-transfer bucket 提供足够的 `/dev/shm` 容量。

### 5. 验证结果

实测平台：4 × Ascend 910B4。

| 指标 | 结果 |
| --- | ---: |
| 连续训练步数 | 100/100 |
| reward 前 10 步均值 | 0.010156 |
| reward 后 10 步均值 | 0.281250 |
| reward 提升 | 0.271094 |
| reward 线性斜率 | 0.00388862/step |
| 每 NPU 平均吞吐 | 258.8282 token/s/NPU |
| 4 NPU 实测总 TPS | 1035.3128 token/s |
| 累计 step 时间 | 4984.8277 s |

100 个 step 连续，reward、actor loss、gradient norm、step time 和 throughput 指标均完整且为有限值。无 GPU 标杆时，结果满足任务要求的 reward 上升与 TPS > 100。

脱敏训练日志:  [training_100step_sanitized.log](https://github.com/user-attachments/files/31128595/training_100step_sanitized.log)

### 6. 代码检查

- 完整 `pre-commit run --all-files`：13/13 passed；
- CISPO objective 与 registry CPU 测试：2 passed；
- example shell 语法检查：passed；
- PR title 与 example filename 检查：passed。

后续将根据 PR review 与 CI 反馈完成调整；PR 合入后在本 Issue 回填最终状态。
